In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [10]:
file = pd.read_csv("titanic.txt", header=None, na_values='?')
file.columns = [
    'pclass', 'survived', 'name', 'sex', 'age', 'sibsp', 'parch', 'ticket',
    'fare', 'cabin', 'embarked', 'boat', 'body', 'home.dest'
]

file.to_csv("titanic_final.csv", index=False)
df = pd.read_csv("titanic_final.csv")
df.head(25)

In [3]:
#zmienne jakosciowe -> zmienne opisowe, dzielace sie na kategorie/grupy
print('Etykiety dla pclass: {}'.format(len(df['pclass'].unique())))
print('Etykiety dla survived: {}'.format(len(df['survived'].unique())))
print('Etykiety dla sex: {}'.format(len(df['sex'].unique())))
print('Etykiety dla sibsp: {}'.format(len(df['sibsp'].unique())))
print('Etykiety dla parch: {}'.format(len(df['parch'].unique())))
print('Etykiety dla cabin: {}'.format(len(df['cabin'].unique())))
print('Etykiety dla embarked: {}'.format(len(df['embarked'].unique())))
print('Etykiety dla boat: {}'.format(len(df['boat'].unique())))
print('Etykiety dla body: {}'.format(len(df['body'].unique())))


Etykiety dla pclass: 3
Etykiety dla survived: 2
Etykiety dla sex: 2
Etykiety dla sibsp: 7
Etykiety dla parch: 8
Etykiety dla cabin: 187
Etykiety dla embarked: 4
Etykiety dla boat: 28
Etykiety dla body: 122


In [4]:
print('Liczba wszystkich pasażerów: {}'.format(len(df)))

Liczba wszystkich pasażerów: 1309


Zmienne 'sex', 'embarked', 'pclass', 'survived', 'sibsp' i 'parch' mają niską kardynalność, więc łatwo je analizować. 'cabin', 'boat' i 'body' mają dużo unikalnych wartości, co może utrudniać pracę. Takie zmienne warto uprościć lub pominąć. 'body' i 'boat' raczej odrzuciłabym, bo ciężko je sensownie zakategoryzować. 'sibsp' i 'parch' można pogrupować na kategorie według liczby osób (ale też bym raczej pominęła).

In [5]:
print('Ilość kabin:',len(np.array(df['cabin'].unique())))

Ilość kabin: 187


In [6]:
df['CabinReduced'] = df['cabin'].astype(str).str[0]
df[['cabin', 'CabinReduced']].head(20)

cabin CabinReduced
0        B5            B
1   C22 C26            C
2   C22 C26            C
3   C22 C26            C
4   C22 C26            C
5       E12            E
6        D7            D
7       A36            A
8      C101            C
9       NaN            n
10  C62 C64            C
11  C62 C64            C
12      B35            B
13      NaN            n
14      A23            A
15      NaN            n
16  B58 B60            B
17  B58 B60            B
18      D15            D
19       C6            C

In [7]:
print('Etykiety dla CabinReduced: {}'.format(len(df['CabinReduced'].unique())))
print('Etykiety dla cabin: {}'.format(len(df['cabin'].unique())))
print('\n')

precent = ((len(df['cabin'].unique()) - len(df['CabinReduced'].unique())) / len(df['cabin'].unique()))*100
print(f'Zredukowano o {precent:.2f}%') 

Etykiety dla CabinReduced: 9
Etykiety dla cabin: 187


Zredukowano o 95.19%


Ilość kabin, która była zajmowana w trakcie podróży jest bardzo wysoka. Utrudnia to dalsze analizy, dlatego w celu ułatwienia pracy, należy w jakiś sposób zredukować kardynalność tej zmiennej. Uważam, że przyjęcie, aby zredukować nazwy kabin tylko do pierwszej litery jest dużym uproszczeniem. Wydaje mi się, że kolejne litery alfabetu oznaczają "piętro", na którym dane kabiny się znajdowały, ewentualnie może jakieś sektory. Taka informacja jest dla nas jakąś wskazówką o ich położeniu i zależności między sobą. Dzięki tej redukcji z 187 etykiet przeszliśmy tylko na 9 co ułatwia dalszą pracę z takimi danymi. Jedyny negatywny skutek jaki mi przychodzi do głowy przy takich zmianach, to utrata bardziej szczegółowych informacji o lokalizacji pasażerów, np: ile osób znajdowało się w konkretnej kabinie.

In [9]:
df = pd.read_csv("titanic.txt", header=None, na_values='?')
df.columns = [
    'pclass', 'survived', 'name', 'sex', 'age', 'sibsp', 'parch', 'ticket',
    'fare', 'cabin', 'embarked', 'boat', 'body', 'home.dest'
]
df['CabinReduced'] = df['cabin'].where(df['cabin'].notna()).astype(str).str[0]
df.to_csv("titanic2.csv", index=False)